cmd/retry-storage.go

```diff
@@ -30,210 +30,276 @@ const (
+func (f *retryStorage) ListDir(volume, path string) (entries []string, err error) {
+	if f.IsOffline() {
+		return entries, errDiskNotFound
+	}
 	entries, err = f.remoteStorage.ListDir(volume, path)
-	if err == errDiskNotFound {
-		err = f.reInit()
-		if err == nil {
-			return f.remoteStorage.ListDir(volume, path)
-		}
+	if f.reInitUponDiskNotFound(err) {
+		entries, err = f.remoteStorage.ListDir(volume, path)
+		return entries, retryToStorageErr(err)
 	}
-	return entries, err
+	return entries, retryToStorageErr(err)
 }
 
 // DeleteFile - a retryable implementation of deleting a file.
-func (f retryStorage) DeleteFile(volume, path string) (err error) {
+func (f *retryStorage) DeleteFile(volume, path string) (err error) {
+	if f.IsOffline() {
+		return errDiskNotFound
+	}
 	err = f.remoteStorage.DeleteFile(volume, path)
-	if err == errDiskNotFound {
-		err = f.reInit()
-		if err == nil {
-			return f.remoteStorage.DeleteFile(volume, path)
-		}
+	if f.reInitUponDiskNotFound(err) {
+		return retryToStorageErr(f.remoteStorage.DeleteFile(volume, path))
 	}
-	return err
+	return retryToStorageErr(err)
 }
 
 // RenameFile - a retryable implementation of renaming a file.
-func (f retryStorage) RenameFile(srcVolume, srcPath, dstVolume, dstPath string) (err error) {
+func (f *retryStorage) RenameFile(srcVolume, srcPath, dstVolume, dstPath string) (err error) {
+	if f.IsOffline() {
+		return errDiskNotFound
+	}
 	err = f.remoteStorage.RenameFile(srcVolume, srcPath, dstVolume, dstPath)
-	if err == errDiskNotFound {
-		err = f.reInit()
-		if err == nil {
-			return f.remoteStorage.RenameFile(srcVolume, srcPath, dstVolume, dstPath)
-		}
+	if f.reInitUponDiskNotFound(err) {
+		return retryToStorageErr(f.remoteStorage.RenameFile(srcVolume, srcPath, dstVolume, dstPath))
 	}
-	return err
+	return retryToStorageErr(err)
+}
+
+// Try to reinitialize the connection when we have some form of DiskNotFound error
+func (f *retryStorage) reInitUponDiskNotFound(err error) bool {
+	if err == errDiskNotFound || err == errDiskNotFoundFromNetError || err == errDiskNotFoundFromRPCShutdown {
+		return f.reInit(err) == nil
+	}
+	return false
 }
 
 // Connect and attempt to load the format from a disconnected node,
 // attempts three times before giving up.
-func (f retryStorage) reInit() (err error) {
+func (f *retryStorage) reInit(e error) (err error) {
+
+	// Only after initialization and minimum of one interval
+	// has passed (to prevent marking a node as offline right
+	// after initialization), check whether node has gone offline
+	if f.maxRetryAttempts == globalStorageRetryThreshold &&
+		UTCNow().Sub(f.offlineTimestamp) >= globalStorageHealthCheckInterval {
+		if e == errDiskNotFoundFromNetError { // Make node offline due to network error
+			f.offline = true // Marking node offline
+			f.offlineTimestamp = UTCNow()
+			return errDiskNotFound
+		} else if e == errDiskNotFoundFromRPCShutdown {
```

Comment:
Is this for explicitness or why is there a no-op `e == errDiskNotFoundFromRPCShutdown` check?